### PR TITLE
Delta housekeeping notebooks

### PR DESCRIPTION
**Pull request by edurdevic**

### CODE REVIEW

File: discoverx/delta_housekeeping.py [hunk 1]
```diff
@@ -0,0 +1,368 @@
...
+        self.tables_not_vacuumed_legend = "Tables that are never VACUUM'ed"
+        self.tables_not_optimized_last_days = "Tables that are not OPTIMIZED often enough"
+        self.tables_not_vacuumed_last_days = "Tables that are not VACUUM'ed often enough"
+        self.tables_optimized_too_freq = "Tables that are OPTIMIZED too often"
+        self.tables_vacuumed_too_freq = "Tables that are VACUUM'ed too often"
+        self.tables_do_not_need_optimize = "Tables that are too small to be OPTIMIZED"
+        self.tables_to_analyze = "Tables that need more analysis (small_files)"
+        self.tables_zorder_not_effective = "Tables for which ZORDER is not being effective"
+
+    def stats(self) -> pd.DataFrame:
```
Comment (edurdevic) [inline discoverx/delta_housekeeping.py:237]: Please remove this method. 
In order to keep the API simple, there should be only one method returning a dataframe, and that method should be `apply` like in `DataExplorerActions`. Please change the implementation of apply to include the diagnostics information as extra boolean columns, see comment below.

File: discoverx/delta_housekeeping.py [hunk 2]
```diff
@@ -0,0 +1,368 @@
...
+            self.tables_vacuumed_too_freq,
+            self.tables_do_not_need_optimize,
+            self.tables_to_analyze,
+            self.tables_zorder_not_effective,
+        ]):
+            if not df.empty:
+                out.append({legend: df})
+        return out
+
+    def to_html(self):
```
[the remark this answers is not in the document]
Comment (edurdevic) [reply discoverx/delta_housekeeping.py:352]: Display should display the full output dataset.
You could use the explain() method to provide action guidance fo users after the analysis is done.

File: examples/exec_delta_housekeeping.py [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+# DBTITLE 1,Run the discoverx DeltaHousekeeping operation -generates an output object you can apply operations to
+output = (
+  dx.from_tables("lorenzorubi.*.*")
+  .delta_housekeeping()
+)
+
+# COMMAND ----------
+
+# DBTITLE 1,Generate a pandas dataframe with stats per table
+display(output.stats())
```
Comment (edurdevic) [inline examples/exec_delta_housekeeping.py:29]: please refactor to `output.display()`

File: examples/exec_delta_housekeeping.py [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+# COMMAND ----------
+
+# DBTITLE 1,apply() operation generates a list of dictionaries (if you need to postprocess the output)
```
Comment (edurdevic) [inline examples/exec_delta_housekeeping.py:33]: Apply returns a spark `DataFrame` in `DataExplorerActions`. Please use the same pattern

File: examples/exec_delta_housekeeping.py [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+result = output.apply()
+
+# COMMAND ----------
+
+for r in result:
```
Comment (edurdevic) [inline examples/exec_delta_housekeeping.py:38]: This should be covered by `result.explain()`, right?

File: examples/exec_delta_housekeeping.py [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  print(list(r.keys())[0])
+  display(list(r.values())[0])
+
+# COMMAND ----------
+
+# DBTITLE 1,to_html() outputs the DeltaHousekeeping recommendations
+displayHTML(output.to_html())
```
Comment (edurdevic) [inline examples/exec_delta_housekeeping.py:45]: Please refactor this with `result.explain()`

File: discoverx/delta_housekeeping.py [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+        - DESCRIBE HISTORY
+        """
+        dd_list = []
+        statements = []
+        errors = []
+
+        if not isinstance(table_info_list, Iterable):
+            table_info_list = [table_info_list]
+
+        for table_info in table_info_list:
```
Comment (edurdevic) [inline discoverx/delta_housekeeping.py:119]: You are not using the `map` function, this is not an iterable any more.
Please simplify this part by removing the for loop.

**Review by edurdevic — COMMENTED** [review]
@lorenzorubi-db Nice work!
I added a number of comments, they are mostly intended to align this functionality with the rest of the library.
This looks very promising!

**Comment by lorenzorubi-db** [comment]
@edurdevic thanks for your comments, I'd say that all of them are covered now, please let me know if you find anything missing or new ideas come up

note that I've removed the implementation of `map_chunked` but will create another PR soon

**Comment by lorenzorubi-db** [comment]
@edurdevic thanks for your comments, I'd say that all of them are covered now, please let me know if you find anything missing or new ideas come up

note that I've removed the implementation of `map_chunked` and added it into #99

File: examples/exec_delta_housekeeping.py [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+
+from discoverx import DX
+
+dx = DX()
+
+# COMMAND ----------
+
+# DBTITLE 1,Run the discoverx DeltaHousekeeping operation -generates an output object you can apply operations to
+output = (
+  dx.from_tables("lorenzorubi.*.*")
```
Comment (edurdevic) [inline examples/exec_delta_housekeeping.py:25]: Uh, I missed this.
Could you please move "lorenzorubi.*.*" to a widget, and if you want replace it with another example catalog name?

File: discoverx/delta_housekeeping.py [hunk 9]
```diff
@@ -0,0 +1,378 @@
...
+        Provides
+        - table stats (size and number of files)
+        - timestamp for last & second last OPTIMIZE
+        - stats of OPTIMIZE (including ZORDER)
+        - timestamp for last & second last VACUUM
+
+        returns a pandas DataFrame, and converts Spark internal dfs to pandas as soon as they are manageable
+        the reason being that DESCRIBE HISTORY / DESCRIBE DETAIL cannot be cached
+
+        TODO reconsider if it is better outside of the class
```
Comment (edurdevic) [inline discoverx/delta_housekeeping.py:39]: Please remove this TODO

File: discoverx/delta_housekeeping.py [hunk 10]
```diff
@@ -0,0 +1,378 @@
...
+    def save_as_table(
+        result: DataFrame,
+        housekeeping_table_name: str,
+    ):
+        """
+        Static method to store intermediate results of the scan operation into Delta
+        Would make sense only if using map_chunked from the `DataExplorer` object
+        (otherwise tables are writen one by one into Delta with overhead)
+
+        TODO create function in `DataExplorer` that uses this for a chunked
```
Comment (edurdevic) [inline discoverx/delta_housekeeping.py:108]: Please remove the TODO

File: tests/unit/delta_housekeeping_actions_test.py [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+    res = dha.generate_recommendations()
+    assert len(res) == 6
+    need_optimize = [item for item in res if (list(item.keys())[0] == dha.tables_not_optimized_legend)]
+    assert len(need_optimize) == 1
+    need_optimize_df = list(need_optimize[0].values())[0]
+    pd.testing.assert_frame_equal(
+        need_optimize_df.reset_index().loc[:, ["catalog", "database", "tableName"]],
+        expected_need_optimize.loc[:, ["catalog", "database", "tableName"]],
+    )
+    # TODO complete all the tests
```
Comment (edurdevic) [inline tests/unit/delta_housekeeping_actions_test.py:37]: Please remove the TODO

File: tests/unit/delta_housekeeping_test.py [hunk 12]
```diff
@@ -0,0 +1,98 @@
...
+    pd.testing.assert_frame_equal(
+        out.reset_index(),
+        dd_click_sales.reset_index(),
+    )
+
+
+def test_process_describe_history_empty_history(spark, dd_click_sales, dh_click_sales):
+    dh = DeltaHousekeeping(spark)
+    describe_detail_df = spark.createDataFrame(dd_click_sales)
+    describe_history_df = spark.createDataFrame(dh_click_sales)
```
Comment (edurdevic) [inline tests/unit/delta_housekeeping_test.py:91]: NIT: I like to define the DFs inline inside the tests, in order to make the tests more readable.
But that's for the next time :)

**Review by edurdevic — COMMENTED** [review]
Thank you for your contribution @lorenzorubi-db!

**Comment by edurdevic** [comment]
Closing this PR so that @lorenzorubi-db can re-open from his account

**Pull request closed without merging**